rts/Lua/LuaImage.cpp

```diff
@@ -0,0 +1,538 @@
+				lua_pushinteger(L, data[i]);
+			}
+		} break;
+		case GL_FLOAT: {
+			const float* data = reinterpret_cast<float*>(image->bitmap->GetRawMem()) + pixelCoords;
+			for (int i=0; i < image->channels; ++i) {
+				lua_pushnumber(L, data[i]);
+			}
+		} break;
+	}
+	return image->channels;
+}
+
+
+int ReadPixelsRaw(lua_State* L, const LuaImageData* image, int startX, int startY, int endX, int endY, int index)
+{
+	if (startX > image->width || startY > image->height || startX < 0 || startY < 0)
+		return 0;
+
+	if (endX > image->width || endY > image->height || endX < 0 || endY < 0 || endX <= startX || endY <= startY)
+		return 0;
+
+	if (!lua_isfunction(L, index))
+		return 0;
+
+	for (int y=startY; y < endY; ++y) {
+		int linepos = y * image->width;
+		for (int x=startX; x < endX; ++x) {
+			int pixelCoords = (x + linepos) * image->channels;
+			lua_pushvalue(L, index);
+			switch (image->bitmap->dataType) {
+				case GL_UNSIGNED_BYTE: {
+					const uint8_t* data = image->bitmap->GetRawMem() + pixelCoords;
+					for (int i=0; i < image->channels; ++i) {
+						lua_pushinteger(L, data[i]);
+					}
+					if (lua_pcall(L, image->channels, 0, 0)) {
+						luaL_error(L, "ReadPixelsRaw failed");
+					}
+				} break;
+				case GL_UNSIGNED_SHORT: {
+					const uint16_t* data = reinterpret_cast<uint16_t*>(image->bitmap->GetRawMem()) + pixelCoords;
+					for (int i=0; i < image->channels; ++i) {
+						lua_pushinteger(L, data[i]);
+					}
+					if (lua_pcall(L, image->channels, 0, 0)) {
+						luaL_error(L, "ReadPixelsRaw failed");
+					}
+				} break;
+				case GL_FLOAT: {
+					const float* data = reinterpret_cast<float*>(image->bitmap->GetRawMem()) + pixelCoords;
+					for (int i=0; i < image->channels; ++i) {
+						lua_pushnumber(L, data[i]);
+					}
+					if (lua_pcall(L, image->channels, 0, 0)) {
+						luaL_error(L, "ReadPixelsRaw failed");
+					}
+				} break;
+			}
+		}
+	}
+	return image->channels;
+}
+
+
+std::shared_ptr<LuaImageData> LuaImage::LoadImageObject(lua_State* L)
+{
+	std::string filename = luaL_checkstring(L, 1);
+	int channels = LuaImage::ParseFormat(L, 2);
+	int dataType = LuaImage::ParseDataType(L, 3);
+	bool luminance = luaL_optboolean(L, 4, false);
+	auto image = make_shared<LuaImageData>(filename, channels, dataType, luminance);
+	if (image->valid) {
+		return image;
+	}
+	return nullptr;
+}
+
+
+/******************************************************************************/
+/******************************************************************************/
+
+int LuaImage::meta_gc(lua_State* L)
+{
+	if (lua_isnil(L, 1))
+		return 0;
+
+	auto image = std::move(*static_cast<std::shared_ptr<LuaImageData>*>(luaL_checkudata(L, 1, "Image")));
+	image = {};
+
+	return 0;
+}
+
+
+int LuaImage::meta_tostring(lua_State* L)
+{
+	if (lua_isnil(L, 1))
+		return 0;
+
+	auto image = toimage(L, 1);
+	lua_pushstring(L, std::format("Image(\"{}\", 0x{:x})", image->filename, reinterpret_cast<intptr_t>(image)).c_str());
```

Comment:
Synced should not receive the address since it's not deterministic. Perhaps put some fake value there, ot just have a different format that doesn't have an address at all.
```suggestion
	const intptr_t address = CLuaHandle::GetHandleSynced(L)
		? 0xDEADBEEF
		: reinterpret_cast<intptr_t>(image)
	;
	lua_pushstring(L, std::format("Image(\"{}\", 0x{:x})", image->filename, address).c_str());
```

Reply:
Not sure tbh, maybe with debug mode or so, but for general usage it's going to be an ugly detail ppl won't understand.

Also, for synced I think better to do a cleaner format in any case, without any hex at all.